scipy/stats/_multicomp.py

```diff
@@ -0,0 +1,253 @@
+
+    """
+    samples, control, rng = iv_dunnett(
+        samples=samples, control=control, random_state=random_state
+    )
+
+    rho, df, n_group = params_dunnett(
+        samples=samples, control=control
+    )
+
+    statistic = np.array([
+        stats.ttest_ind(
+            obs_, control, alternative=alternative, random_state=rng
+        ).statistic
+        for obs_ in samples
+    ])
+
+    pvalue = pvalue_dunnett(
+        rho=rho, df=df,
+        statistic=abs(statistic), alternative=alternative,
+        rng=rng
+    )
+
+    return DunnettResult(
+        statistic=statistic, pvalue=pvalue,
+    )
+
+
+def iv_dunnett(
+    samples: npt.ArrayLike, control: npt.ArrayLike, random_state: SeedType
+) -> tuple[npt.ArrayLike, np.ndarray, SeedType]:
+    """Input validation for Dunnett's test."""
+    rng = check_random_state(random_state)
+
+    ndim_msg = "Control and samples groups must be 1D arrays"
+    n_obs_msg = "Control and samples groups must have at least 1 observation"
+
+    control = np.asarray(control)
+    samples = [np.asarray(sample) for sample in samples]
+
+    # samples checks
+    for sample in (samples + [control]):
+        sample = np.asarray(sample)
+        if sample.ndim > 1:
+            raise ValueError(ndim_msg)
+
+        if len(sample) < 1:
+            raise ValueError(n_obs_msg)
+
+    return samples, control, rng
+
+
+def params_dunnett(
+    samples: npt.ArrayLike, control: npt.ArrayLike
+) -> tuple[np.ndarray, int, int]:
+    """Specific parameters for Dunnett's test.
+
+    Covariance matrix depends on the number of observations in each group:
+
+    - All groups are equals (including the control), ``rho_ij=0.5`` except for
+      the diagonal which is 1.
+    - All groups but the control are equal, balanced design.
+    - Groups are not equal, unbalanced design.
+
+    Degree of freedom is the number of observations minus the number of groups
+    including the control.
+    """
+    n_n_obs = np.array([len(obs_) for obs_ in samples])
+
+    # From Dunnett1955 p. 1100 d.f. = (sum N)-(p+1)
+    n_obs = n_n_obs.sum()
+    n_control = len(control)
+    n = n_obs + n_control
+    n_groups = len(samples)
+    df = n - n_groups - 1
+
+    # rho_ij = 1/sqrt((N0/Ni+1)(N0/Nj+1))
+    rho = n_control/n_n_obs + 1
+    rho = 1/np.sqrt(rho[:, None] * rho[None, :])
+    np.fill_diagonal(rho, 1)
+
+    return rho, df, n_groups
+
+
+def pvalue_dunnett(
+    rho: np.ndarray, df: int, statistic: np.ndarray,
+    alternative: Literal['two-sided', 'less', 'greater'],
+    rng: SeedType = None
+) -> np.ndarray:
+    """pvalue from Dunnett critical value.
+
+    Critical values come from the multivariate student-t distribution.
+    """
+    statistic = statistic.reshape(-1, 1)
+
+    mvt = stats.multivariate_t(shape=rho, df=df, seed=rng)
+    if alternative == "two-sided":
+        pvalue = 1 - mvt.cdf(statistic, lower_limit=-statistic)
+    else:
+        pvalue = 1 - mvt.cdf(statistic)
```

Comment:
I think this just covers `alternative='greater'`. You need to integrate from `statistic` to `np.inf` for `alternative='less'`